File: app/src/main/java/com/example/bookapp/viewModels/ViewModelPost.java
package com.example.bookapp.viewModels;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.example.bookapp.activities.AppUtilities;
import com.example.bookapp.models.Post;
import com.example.dataLayer.dataObjectsToSerialize.SerializePost;
import com.example.dataLayer.repositories.PostRepository;

import java.util.ArrayList;

public class ViewModelPost extends ViewModel {
    private MutableLiveData<ArrayList<Post>> currentFetchedPosts;
    private MutableLiveData<ArrayList<Post>> currentlyDisplayedPosts;

    private MutableLiveData<ArrayList<Post>> myPosts;
    private MutableLiveData<ArrayList<Post>> newFetchedPosts ;

    private PostRepository postRepository;


    public ViewModelPost() {
        super();
        postRepository = PostRepository.getInstance(AppUtilities.getRetrofit());
    }


    public MutableLiveData<ArrayList<Post>> getMyPosts(String userID) {
        if (myPosts == null) {
            //should fetch posts
            myPosts = postRepository.fetchMyPosts(userID);
        }
        return myPosts;
    }

    public MutableLiveData<Post> getPost(long id) {
        return postRepository.fetchPostByID(id);
    }





    public LiveData<ArrayList<Post>> getFirstPagePosts(int page) {

        if (currentFetchedPosts == null) {
            currentFetchedPosts = postRepository.fetchFirstPagePosts();
        }
        return currentFetchedPosts;
    }



    public MutableLiveData<ArrayList<Post>> getFavoritePosts(String userID) {
        if (currentlyDisplayedPosts == null) {
            currentlyDisplayedPosts = postRepository.fetchFavoritePosts(userID);
        }
        return currentlyDisplayedPosts;
    }


    public void addPostToFavorites(Post post, String userID) {
        postRepository.addPostToFavorites(post.getPostID(), userID);
        //if the favoritePosts data has been fetched you need to update the ui
        if (currentlyDisplayedPosts != null && currentlyDisplayedPosts.getValue() != null) {
            ArrayList<Post> newData = new ArrayList<>(currentlyDisplayedPosts.getValue());
            post.setFavorite(true);
            newData.add(post);
            currentlyDisplayedPosts.setValue(newData);
        }
    }

    public void deletePostFromFavorites(Post post, String userID) {
        postRepository.deletePostFromFavorites(post.getPostID(), userID);
        if (currentlyDisplayedPosts != null && currentlyDisplayedPosts.getValue() != null) {
            ArrayList<Post> newData = new ArrayList<>(currentlyDisplayedPosts.getValue());
            newData.remove(post);
            //notify the observers
            currentlyDisplayedPosts.setValue(newData);
        }
    }

    public void addPost(@NonNull SerializePost serializePost) {
        postRepository.insertPost(serializePost);
    }

    public void fetchNewPosts() {
        postRepository.fetchNewPosts();
    }


    public MutableLiveData<ArrayList<Post>> getNewFetchedPosts() {
        return postRepository.getNewFetchedPosts();
    }
}

File: app/src/main/java/com/example/bookapp/viewModels/ViewModelPost.kt
package com.example.bookapp.viewModels

import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.example.bookapp.models.Post
import com.example.dataLayer.models.SerializeImage
import com.example.dataLayer.models.UserWithFavoritePosts
import com.example.dataLayer.models.serialization.SerializePost
import com.example.dataLayer.repositories.OperationStatus
import com.example.dataLayer.repositories.PostRepository
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.launch
import javax.inject.Inject

@InternalCoroutinesApi
class ViewModelPost @ViewModelInject constructor(
        private val postRepository: PostRepository
) : ViewModel() {

    
    fun getFavoritePosts(): LiveData<UserWithFavoritePosts> = postRepository.favoritePosts


    val userPosts by lazy {
        postRepository.fetchMyPosts()
    }


    fun getPostByID(id: String): LiveData<Post> {
        return postRepository.fetchPostByID(id)
    }

    private val config = PagedList.Config.Builder()
            .setPageSize(10)
            .setInitialLoadSizeHint(20)
            .setEnablePlaceholders(true)
            .setPrefetchDistance(5)
            .build()


    val recentPosts by lazy {
        LivePagedListBuilder(postRepository.getPosts(), config)
                .setBoundaryCallback(postRepository.PostRepoBoundaryCallback())
                .build()
    }

    fun addPostToFavorites(post: Post) {
        viewModelScope.launch {
            postRepository.addPostToFavorites(post)
        }
    }

    fun deletePostFromFavorites(post: Post) =
            viewModelScope.launch { postRepository.deletePostFromFavorites(post) }


    fun fetchNewPosts() = viewModelScope.launch { postRepository.fetchInitialPosts() }

    fun uploadImage(serializeImage: SerializeImage) = postRepository.uploadImage(serializeImage)

    fun uploadPost(post: SerializePost): LiveData<OperationStatus> = postRepository.uploadPost(post)


}